helpers/oci-validate.sh

```diff
@@ -23,85 +23,131 @@ jq -L"$BASHBREW_META_SCRIPTS" --slurp '
 	| empty
 ' oci-layout
 
-# TODO this is all rubbish; it needs more thought (the jq functions it invokes are pretty solid now though)
+# TODO (recursively?) validate subject descriptors in here somewhere 🤔
 
 descriptor() {
 	local file="$1"; shift # "blobs/sha256/xxx"
-	echo "blob: $file"
-	local digest="$1"; shift # "sha256:xxx"
-	local size="$1"; shift # "123"
-	local algo="${digest%%:*}" # sha256
-	local hash="${digest#$algo:}" # xxx
-	local diskSize
-	[ "$algo" = 'sha256' ] # TODO error message
-	diskSize="$(stat --dereference --format '%s' "$file")"
-	[ "$size" = "$diskSize" ] # TODO error message
-	"${algo}sum" <<<"$hash *$file" --check --quiet --strict -
+	local desc; desc="$(cat)"
+	local shell
+	shell="$(jq <<<"$desc" -L"$BASHBREW_META_SCRIPTS" --slurp --raw-output '
+		include "validate";
+		include "oci";
+		validate_one
+		| validate_oci_descriptor
+		| (
+			@sh "local algo=\(
+				.digest
+				| split(":")[0]
+				| validate_IN(.; "sha256", "sha512") # TODO more algorithms? need more tools on the host
+			)",
+
+			@sh "local data=\(
+				if has("data") then
+					.data
+				else " " end # empty string is valid base64 (which we should validate), but spaces are not, so we can use a single space to detect "data not set"
+			)",
+
+			empty
+		)
+	')"
+	eval "$shell"
+	local digest size dataDigest= dataSize=
+	digest="$("${algo}sum" "$file" | cut -d' ' -f1)"
+	digest="$algo:$digest"
+	size="$(stat --dereference --format '%s' "$file")"
+	if [ "$data" != ' ' ]; then
+		dataDigest="$(base64 <<<"$data" -d | "${algo}sum" | cut -d' ' -f1)"
+		dataDigest="$algo:$dataDigest"
+		dataSize="$(base64 <<<"$data" -d | wc --bytes)"
+		# TODO *technically* we could get clever here and pass `base64 -d` to something like `tee >(wc --bytes) >(dig="$(sha256sum | cut -d' ' -f1)" && echo "sha256:$dig" && false) > /dev/null` to avoid parsing the base64 twice, but then failure cases are less likely to be caught, so it's safer to simply redecode (and we can't decode into a variable because this might be binary data *and* bash will do newline munging in both directions)
+	fi
```

Comment:
Should this be able to handle a descriptor that only has `data` and not also a file on disk? Or is that not a valid OCI layout?

Reply:
This is such an (unintentionally) brutal comment :joy: :sob:

You're absolutely right, and making me question whether I like this interface again.

Implementations worried about portability should absolutely have both `data` _and_ the blob in the `blobs/` folder, but it's not technically invalid not to, and I'd love to support that. :thinking: